File: source/dotnet/Library/AdaptiveCards/AdaptiveMedia.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
using System.Collections.Generic;
using System.ComponentModel;
using System.Xml.Serialization;
using System.Text.Json.Serialization;

namespace AdaptiveCards
{
    /// <summary>
    /// Represents the Media element.
    /// </summary>
    [XmlType(TypeName = TypeName)]
    public class AdaptiveMedia : AdaptiveElement
    {
        /// <inheritdoc />
        public const string TypeName = "Media";

        /// <inheritdoc />
        [XmlIgnore]
        public override string Type { get; set; } = TypeName;

        /// <summary>
        /// A collection of source from which to retrieve the media.
        /// </summary>
        [JsonRequired]
        [XmlElement(Type = typeof(AdaptiveMediaSource), ElementName = "MediaSource")]
        public List<AdaptiveMediaSource> Sources { get; set; } = new List<AdaptiveMediaSource>();

        /// <summary>
        /// URL for the poster image to show for this media element.
        /// </summary>
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        [XmlAttribute]
        [DefaultValue(null)]
        public string Poster { get; set; }

        /// <summary>
        /// Alternate text to display for this media element.
        /// </summary>
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        [XmlAttribute]
        [DefaultValue(null)]
        public string AltText { get; set; }

        /// <summary>
        /// A collection of captions.
        /// </summary>
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        [XmlElement(Type = typeof(AdaptiveCaptionSource), ElementName = "CaptionSource")]
        [DefaultValue(null)]
        public List<AdaptiveCaptionSource> CaptionSources { get; set; } = new List<AdaptiveCaptionSource>();

    }
}

File: source/uwp/SharedObjectModel/lib/AdaptiveMedia.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
#include "pch.h"
#include "AdaptiveMedia.h"
#include "AdaptiveMedia.g.cpp"
#include "AdaptiveMediaSource.h"
#include "AdaptiveCaptionSource.h"

namespace winrt::AdaptiveCards::ObjectModel::Xaml_OM::implementation
{
    AdaptiveMedia::AdaptiveMedia(const std::shared_ptr<::AdaptiveCards::Media>& sharedMedia)
    {
        Poster = UTF8ToHString(sharedMedia->GetPoster());
        AltText = UTF8ToHString(sharedMedia->GetAltText());
        Sources = GenerateVectorProjection<winrt::AdaptiveCards::ObjectModel::Xaml_OM::implementation::AdaptiveMediaSource>(sharedMedia->GetSources());
        CaptionSources = GenerateVectorProjection<winrt::AdaptiveCards::ObjectModel::Xaml_OM::implementation::AdaptiveCaptionSource>(sharedMedia->GetCaptionSources());
        InitializeBaseElement(sharedMedia);
    }

    std::shared_ptr<::AdaptiveCards::BaseCardElement> AdaptiveMedia::GetSharedModel()
    {
        auto media = std::make_shared<::AdaptiveCards::Media>();
        CopySharedElementProperties(*media);
        media->SetPoster(HStringToUTF8(Poster));
        media->SetAltText(HStringToUTF8(AltText));
        media->GetSources() =
            GenerateSharedVector<winrt::AdaptiveCards::ObjectModel::Xaml_OM::implementation::AdaptiveMediaSource, ::AdaptiveCards::MediaSource>(Sources.get());
        media->GetCaptionSources() =
            GenerateSharedVector<winrt::AdaptiveCards::ObjectModel::Xaml_OM::implementation::AdaptiveCaptionSource, ::AdaptiveCards::CaptionSource>(CaptionSources.get());

        return media;
    }
}
